{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Cluster and Family Label\n",
    "\n",
    "Code for clustering proteins and nucleic acids by sequence and code for labeling protein families."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import shutil\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Paths and Directories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Directory of all datasets.\n",
    "dataset_directory = \"/home/akubaney/projects/na_mpnn/data/datasets\"\n",
    "\n",
    "# List of datasets to family label and cluster.\n",
    "dataset_names = [\"rcsb_cif_na\", \"rf2na_distillation_cis_bp\", \"rf2na_distillation_transfac\"]\n",
    "\n",
    "# Directory for protein family labeling.\n",
    "protein_family_labeling_directory = \"./protein_family_labeling\"\n",
    "num_jobs_for_family_labeling = 1000\n",
    "\n",
    "# Directory for clustering.\n",
    "clustering_directory = \"./clustering\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Helper Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_text_file(path):\n",
    "    with open(path, mode = \"rt\") as f:\n",
    "        return f.read()\n",
    "\n",
    "def write_text_file(path, contents):\n",
    "    with open(path, mode = \"wt\") as f:\n",
    "        f.write(contents)\n",
    "\n",
    "def read_fasta_file(path):\n",
    "    fasta_text = read_text_file(path).strip()\n",
    "    fasta_entries = fasta_text[1:].split(\"\\n>\")\n",
    "    id_and_sequence_pairs = []\n",
    "    for fasta_entry in fasta_entries:\n",
    "        header, sequence = fasta_entry.strip().split(\"\\n\")\n",
    "        id = header.strip()\n",
    "        id_and_sequence_pairs.append((id, sequence))\n",
    "    return id_and_sequence_pairs\n",
    "\n",
    "def write_fasta_file(path, id_and_sequence_pairs):\n",
    "    fasta_entries = []\n",
    "    for id, sequence in id_and_sequence_pairs:\n",
    "        fasta_entries.append(f\">{id}\\n{sequence}\")\n",
    "    fasta_text = \"\\n\".join(fasta_entries)\n",
    "    write_text_file(path, fasta_text)\n",
    "\n",
    "def read_cdhit_cluster_file(path):\n",
    "    clusters_text = read_text_file(path).strip()\n",
    "    cluster_entries = clusters_text[1:].split(\"\\n>\")\n",
    "    clusters = dict()\n",
    "    for cluster_entry in cluster_entries:\n",
    "        cluster_entry_lines = cluster_entry.strip().split(\"\\n\")\n",
    "\n",
    "        # Extract the cluster id from the header.\n",
    "        cluster_header_line = cluster_entry_lines[0]\n",
    "        cluster_id = int(cluster_header_line.strip().split(\" \")[1])\n",
    "\n",
    "        # Extract the cluster members.\n",
    "        cluster_member_lines = cluster_entry_lines[1:]\n",
    "        cluster_members = []\n",
    "        for cluster_member_line in cluster_member_lines:\n",
    "            member_length, member_entry = cluster_member_line.strip().split(\", >\")\n",
    "            member_id, _ = member_entry.split(\"...\")\n",
    "            cluster_members.append(member_id)\n",
    "\n",
    "        clusters[cluster_id] = cluster_members\n",
    "    \n",
    "    return clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_cluster_df_from_sequences_and_clusters(fasta_path, \n",
    "                                               clusters_path,\n",
    "                                               cluster_id_column_name):\n",
    "    # Load the original fasta file.\n",
    "    id_and_sequence_pairs = read_fasta_file(fasta_path)\n",
    "    print(f\"Initial number of sequences: {len(id_and_sequence_pairs)}\")\n",
    "\n",
    "    # Create a dataframe for the sequences and their cluster ids.\n",
    "    cluster_data_dict = {\"id\": [], \"sequence\": []}\n",
    "\n",
    "    for id, sequence in id_and_sequence_pairs:\n",
    "        cluster_data_dict[\"id\"].append(id)\n",
    "        cluster_data_dict[\"sequence\"].append(sequence)\n",
    "\n",
    "    cluster_df = pd.DataFrame(cluster_data_dict)\n",
    "\n",
    "    # Load the clusters.\n",
    "    clusters = read_cdhit_cluster_file(clusters_path)\n",
    "    print(f\"Number of clusters: {len(clusters)}\")\n",
    "\n",
    "    # Count for sequences that failed to cluster.\n",
    "    all_sequence_ids = set(cluster_df[\"id\"])\n",
    "\n",
    "    clustered_sequence_ids = set()\n",
    "    for cluster_id in clusters:\n",
    "        clustered_sequence_ids.update(clusters[cluster_id])\n",
    "\n",
    "    unclustered_sequence_ids = all_sequence_ids.difference(clustered_sequence_ids)\n",
    "    print(f\"Number of unclustered sequences: {len(unclustered_sequence_ids)}\")\n",
    "\n",
    "    # Create a mapping of sequence id to cluster id.\n",
    "    sequence_id_to_cluster_id = dict()\n",
    "    for cluster_id in clusters:\n",
    "        for sequence_id in clusters[cluster_id]:\n",
    "            assert(sequence_id not in sequence_id_to_cluster_id)\n",
    "            sequence_id_to_cluster_id[sequence_id] = cluster_id\n",
    "    print(f\"Number of sequence id to cluster id mappings: {len(sequence_id_to_cluster_id)}\")\n",
    "\n",
    "    # Add the cluster id to the dataframe.\n",
    "    cluster_df[cluster_id_column_name] = cluster_df[\"id\"].map(sequence_id_to_cluster_id)\n",
    "\n",
    "    # Drop the sequences that failed to cluster.\n",
    "    cluster_df = cluster_df.dropna(subset = [cluster_id_column_name])\n",
    "\n",
    "    # Make the cluster id an integer.\n",
    "    cluster_df[cluster_id_column_name] = cluster_df[cluster_id_column_name].astype(int)\n",
    "\n",
    "    return cluster_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Family Label All Protein Sequences"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create the directory for family labeling.\n",
    "if os.path.exists(protein_family_labeling_directory):\n",
    "    shutil.rmtree(protein_family_labeling_directory)\n",
    "os.makedirs(protein_family_labeling_directory)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Gather All Protein Sequnces"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_protein_sequences = set()\n",
    "# For every dataset.\n",
    "for dataset_name in dataset_names:\n",
    "    preprocessing_output = os.path.join(dataset_directory, \n",
    "                                        dataset_name, \n",
    "                                        \"preprocessing_output.csv\")\n",
    "    preprocessing_output_df = pd.read_csv(preprocessing_output)\n",
    "\n",
    "    # For every structure in the dataset.\n",
    "    for sequences_path in preprocessing_output_df[\"sequences_path\"]:\n",
    "        sequences_df = pd.read_csv(sequences_path)\n",
    "\n",
    "        # For every chain sequence in the structure.\n",
    "        for chain_type, sequence in zip(sequences_df[\"chain_type\"], sequences_df[\"sequence\"]):\n",
    "            # If the chain is a protein, add the sequence.\n",
    "            if chain_type == \"polypeptide(L)\":\n",
    "                all_protein_sequences.add(sequence)\n",
    "\n",
    "print(len(all_protein_sequences))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Save the Gathered Sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_protein_sequences_path = os.path.join(protein_family_labeling_directory, \"all_protein_sequences.fa\")\n",
    "write_fasta_file(all_protein_sequences_path, enumerate(all_protein_sequences))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run InterScanPro and Process Results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make the directory for the fasta splits.\n",
    "fasta_splits_directory = os.path.join(protein_family_labeling_directory, \"fasta_splits\")\n",
    "os.makedirs(fasta_splits_directory, exist_ok = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Split the all protein sequence fasta into smaller fastas.\n",
    "all_protein_sequences_path = os.path.join(protein_family_labeling_directory, \"all_protein_sequences.fa\")\n",
    "id_and_sequence_pairs = read_fasta_file(all_protein_sequences_path)\n",
    "num_sequences_per_job = (len(all_protein_sequences) + num_jobs_for_family_labeling - 1) // num_jobs_for_family_labeling\n",
    "for job_index in range(num_jobs_for_family_labeling):\n",
    "    start_index = job_index * num_sequences_per_job\n",
    "    end_index = min((job_index + 1) * num_sequences_per_job, len(all_protein_sequences))\n",
    "\n",
    "    split_fasta_path = os.path.join(fasta_splits_directory, \n",
    "                                    f\"all_protein_sequences_{job_index}.fa\")\n",
    "    write_fasta_file(split_fasta_path, id_and_sequence_pairs[start_index : end_index])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make the directory for the family labeling outputs.\n",
    "family_label_output_directory = os.path.join(protein_family_labeling_directory, \"out\")\n",
    "os.makedirs(family_label_output_directory, exist_ok = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Run the following:\n",
    "\n",
    "```\n",
    "cd /home/akubaney/projects/na_mpnn/data\n",
    "\n",
    "fasta_splits_directory=\"./protein_family_labeling/fasta_splits\"\n",
    "family_label_output_directory=\"./protein_family_labeling/out\"\n",
    "family_label_tmp_path=\"./protein_family_labeling/family_label_tmp.out\"\n",
    "\n",
    "rm $family_label_tmp_path\n",
    "\n",
    "sbatch --output=$family_label_tmp_path --array=0-999 ./family_label.sh $fasta_splits_directory $family_label_output_directory\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "interproscan_column_names = [\n",
    "    \"protein_accession\", \n",
    "    \"sequence_md5\", \n",
    "    \"sequence_length\", \n",
    "    \"analysis\", \n",
    "    \"signature_accession\", \n",
    "    \"signature_description\", \n",
    "    \"start_location\", \n",
    "    \"stop_location\", \n",
    "    \"score\", \n",
    "    \"status\", \n",
    "    \"date\", \n",
    "    \"interpro_accession\", \n",
    "    \"interpro_description\", \n",
    "    \"go_annotations\",\n",
    "    \"pathway_annotations\"\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the protein family labels.\n",
    "protein_family_label_dfs = []\n",
    "for protein_family_label_csv_name in os.listdir(family_label_output_directory):\n",
    "    protein_family_label_csv_path = os.path.join(family_label_output_directory, \n",
    "                                                 protein_family_label_csv_name)\n",
    "    protein_family_label_df = pd.read_csv(protein_family_label_csv_path, \n",
    "                                          names = interproscan_column_names, \n",
    "                                          sep = \"\\t\")\n",
    "\n",
    "    protein_family_label_dfs.append(protein_family_label_df)\n",
    "\n",
    "# Combine the results.\n",
    "all_protein_family_label_df = pd.concat(protein_family_label_dfs, ignore_index = True)\n",
    "\n",
    "# Sort by temporary id.\n",
    "all_protein_family_label_df = all_protein_family_label_df.sort_values(by = \"protein_accession\", \n",
    "                                                                      ignore_index = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the original fasta file.\n",
    "all_protein_sequences_path = os.path.join(protein_family_labeling_directory, \"all_protein_sequences.fa\")\n",
    "id_and_sequence_pairs = read_fasta_file(all_protein_sequences_path)\n",
    "\n",
    "# Turn the temporary id back into the sequence.\n",
    "temp_id_to_sequence = dict(id_and_sequence_pairs)\n",
    "all_protein_family_label_df[\"sequence\"] = all_protein_family_label_df[\"protein_accession\"].astype(str).map(temp_id_to_sequence)\n",
    "\n",
    "# Drop the temporary id.\n",
    "all_protein_family_label_df = all_protein_family_label_df.drop(columns = [\"protein_accession\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_protein_family_label_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the results.\n",
    "protein_family_label_output_path = os.path.join(protein_family_labeling_directory, \"all_protein_family_labels.csv\")\n",
    "all_protein_family_label_df.to_csv(protein_family_label_output_path, index = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Clustering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create the directory for clustering.\n",
    "if os.path.exists(clustering_directory):\n",
    "    shutil.rmtree(clustering_directory)\n",
    "os.makedirs(clustering_directory)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Cluster All Protein Sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create the directory for the all protein sequences clustering.\n",
    "all_protein_sequences_clustering_directory = os.path.join(clustering_directory, \"all_protein_sequences\")\n",
    "if os.path.exists(all_protein_sequences_clustering_directory):\n",
    "    shutil.rmtree(all_protein_sequences_clustering_directory)\n",
    "os.makedirs(all_protein_sequences_clustering_directory)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run CD-HIT on the Protein Sequences"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```\n",
    "cd /home/akubaney/projects/na_mpnn/data\n",
    "\n",
    "input_fasta_path=\"./protein_family_labeling/all_protein_sequences.fa\"\n",
    "output_path=\"./clustering/all_protein_sequences/all_protein_sequences\"\n",
    "sequence_similarity_cutoff=0.4\n",
    "length_similarity_cutoff=0.9\n",
    "word_size=2\n",
    "\n",
    "/home/akubaney/software/cd-hit-v4.8.1-2019-0228/cd-hit \\\n",
    "    -i $input_fasta_path \\\n",
    "    -o $output_path \\\n",
    "    -c $sequence_similarity_cutoff \\\n",
    "    -n $word_size \\\n",
    "    -d 0 \\\n",
    "    -M 16000 \\\n",
    "    -T 0 \\\n",
    "    -aL $length_similarity_cutoff \\\n",
    "    -aS $length_similarity_cutoff\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Compute and Save Chain Sequence Clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# The fasta path for all protein sequences.\n",
    "all_protein_sequences_path = os.path.join(protein_family_labeling_directory, \"all_protein_sequences.fa\")\n",
    "\n",
    "# The output path for the cd-hit clustering.\n",
    "protein_clusters_path = os.path.join(all_protein_sequences_clustering_directory, \"all_protein_sequences.clstr\")\n",
    "\n",
    "# Create a dataframe that contains the protein sequences and their cluster ids.\n",
    "protein_cluster_df = get_cluster_df_from_sequences_and_clusters(all_protein_sequences_path, \n",
    "                                                                protein_clusters_path,\n",
    "                                                                \"protein_chain_cluster_id\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "protein_cluster_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the results.\n",
    "protein_cluster_out_path = os.path.join(all_protein_sequences_clustering_directory, \"all_protein_sequences_clusters.csv\")\n",
    "protein_cluster_df.to_csv(protein_cluster_out_path, index = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Cluster All Nucleic Acid Sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create the directory for the all nucleic acid sequences clustering.\n",
    "all_nucleic_acid_sequences_clustering_directory = os.path.join(clustering_directory, \"all_nucleic_acid_sequences\")\n",
    "if os.path.exists(all_nucleic_acid_sequences_clustering_directory):\n",
    "    shutil.rmtree(all_nucleic_acid_sequences_clustering_directory)\n",
    "os.makedirs(all_nucleic_acid_sequences_clustering_directory)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Gather All Nucleic Acid Sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def standardize_na_sequence(sequence):\n",
    "    \"\"\"\n",
    "    Given a nucleic acid sequence string, map U->T, and any residue that is not\n",
    "    A,C,G,T to X.\n",
    "    \"\"\"\n",
    "    nucleic_acid_mapping = {\n",
    "        \"A\": \"A\",\n",
    "        \"C\": \"C\",\n",
    "        \"G\": \"G\",\n",
    "        \"T\": \"T\",\n",
    "        \"U\": \"T\"\n",
    "    }\n",
    "\n",
    "    sequence = \"\".join(list(map(lambda c: nucleic_acid_mapping.get(c, \"X\"), list(sequence))))\n",
    "    return sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_nucleic_acid_sequences = set()\n",
    "all_nucleic_acid_standard_sequences = set()\n",
    "# For every dataset.\n",
    "for dataset_name in dataset_names:\n",
    "    preprocessing_output = os.path.join(dataset_directory, \n",
    "                                        dataset_name, \n",
    "                                        \"preprocessing_output.csv\")\n",
    "    preprocessing_output_df = pd.read_csv(preprocessing_output)\n",
    "\n",
    "    # For every structure in the dataset.\n",
    "    for sequences_path in preprocessing_output_df[\"sequences_path\"]:\n",
    "        sequences_df = pd.read_csv(sequences_path)\n",
    "\n",
    "        # For every chain sequence in the structure.\n",
    "        for chain_type, sequence in zip(sequences_df[\"chain_type\"], sequences_df[\"sequence\"]):\n",
    "            # If the chain is a nucleic acid, add the sequence.\n",
    "            if chain_type in [\"polydeoxyribonucleotide/polyribonucleotide hybrid\", \n",
    "                              \"polydeoxyribonucleotide\", \n",
    "                              \"polyribonucleotide\"]:\n",
    "                all_nucleic_acid_sequences.add(sequence)\n",
    "                all_nucleic_acid_standard_sequences.add(standardize_na_sequence(sequence))\n",
    "\n",
    "print(len(all_nucleic_acid_sequences))\n",
    "print(len(all_nucleic_acid_standard_sequences))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_nucleic_acid_sequences_path = os.path.join(all_nucleic_acid_sequences_clustering_directory, \n",
    "                                               \"all_nucleic_acid_sequences.fa\")\n",
    "write_fasta_file(all_nucleic_acid_sequences_path, enumerate(all_nucleic_acid_sequences))\n",
    "\n",
    "all_nucleic_acid_standard_sequences_path = os.path.join(all_nucleic_acid_sequences_clustering_directory, \n",
    "                                                        \"all_nucleic_acid_standard_sequences.fa\")\n",
    "write_fasta_file(all_nucleic_acid_standard_sequences_path, enumerate(all_nucleic_acid_standard_sequences))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run CD-HIT on the Nucleic Acid Sequences"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```\n",
    "cd /home/akubaney/projects/na_mpnn/data\n",
    "\n",
    "input_fasta_path=\"./clustering/all_nucleic_acid_sequences/all_nucleic_acid_standard_sequences.fa\"\n",
    "output_path=\"./clustering/all_nucleic_acid_sequences/all_nucleic_acid_standard_sequences\"\n",
    "sequence_similarity_cutoff=0.8\n",
    "length_similarity_cutoff=0.9\n",
    "word_size=4\n",
    "length_limit=4\n",
    "\n",
    "/home/akubaney/software/cd-hit-v4.8.1-2019-0228/cd-hit-est \\\n",
    "    -i $input_fasta_path \\\n",
    "    -o $output_path \\\n",
    "    -c $sequence_similarity_cutoff \\\n",
    "    -n $word_size \\\n",
    "    -d 0 \\\n",
    "    -M 16000 \\\n",
    "    -T 0 \\\n",
    "    -l $length_limit \\\n",
    "    -aL $length_similarity_cutoff \\\n",
    "    -aS $length_similarity_cutoff\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Compute and Save Chain Sequence Clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# The fasta path for all nucleic acid standard sequences.\n",
    "all_nucleic_acid_standard_sequences_path = os.path.join(all_nucleic_acid_sequences_clustering_directory, \n",
    "                                                        \"all_nucleic_acid_standard_sequences.fa\")\n",
    "\n",
    "# The output path for the cd-hit clustering.\n",
    "nucleic_acid_standard_clusters_path = os.path.join(all_nucleic_acid_sequences_clustering_directory, \n",
    "                                                   \"all_nucleic_acid_standard_sequences.clstr\")\n",
    "\n",
    "# Create a dataframe that contains the nucleic acid standard sequences and their cluster ids.\n",
    "nucleic_acid_standard_cluster_df = \\\n",
    "    get_cluster_df_from_sequences_and_clusters(all_nucleic_acid_standard_sequences_path, \n",
    "                                               nucleic_acid_standard_clusters_path,\n",
    "                                               \"nucleic_acid_standard_chain_cluster_id\")\n",
    "standard_sequence_to_cluster_id = dict(zip(nucleic_acid_standard_cluster_df[\"sequence\"], \n",
    "                                           nucleic_acid_standard_cluster_df[\"nucleic_acid_standard_chain_cluster_id\"]))\n",
    "\n",
    "# Create a dataframe that contains the nucleic acid sequences and their standard cluster ids.\n",
    "all_nucleic_acid_sequences_path = os.path.join(all_nucleic_acid_sequences_clustering_directory,\n",
    "                                               \"all_nucleic_acid_sequences.fa\")\n",
    "id_and_sequence_pairs = read_fasta_file(all_nucleic_acid_sequences_path)\n",
    "\n",
    "nucleic_acid_cluster_data_dict = {\"id\": [], \"sequence\": [], \"nucleic_acid_chain_cluster_id\": []}\n",
    "for id, sequence in id_and_sequence_pairs:\n",
    "    standard_sequence = standardize_na_sequence(sequence)\n",
    "    # Since non-clustered sequences were dropped, we need to check this here.\n",
    "    if standard_sequence in standard_sequence_to_cluster_id:\n",
    "        cluster_id = standard_sequence_to_cluster_id[standard_sequence]\n",
    "        nucleic_acid_cluster_data_dict[\"id\"].append(id)\n",
    "        nucleic_acid_cluster_data_dict[\"sequence\"].append(sequence)\n",
    "        nucleic_acid_cluster_data_dict[\"nucleic_acid_chain_cluster_id\"].append(cluster_id)\n",
    "\n",
    "nucleic_acid_cluster_df = pd.DataFrame(nucleic_acid_cluster_data_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "nucleic_acid_cluster_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the results.\n",
    "nucleic_acid_cluster_out_path = os.path.join(all_nucleic_acid_sequences_clustering_directory, \n",
    "                                             \"all_nucleic_acid_sequences_clusters.csv\")\n",
    "nucleic_acid_cluster_df.to_csv(nucleic_acid_cluster_out_path, index = False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
